File: configgen/src/main/scala/com/yahoo/config/codegen/BuilderGenerator.scala
// Copyright 2016 Yahoo Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.config.codegen

import com.yahoo.config.codegen.ReservedWords.{INTERNAL_PREFIX => InternalPrefix}
import JavaClassBuilder.{Indentation, createUniqueSymbol}
import ConfigGenerator.{indentCode, nodeClass, userDataType, boxedDataType}
import com.yahoo.config.codegen.LeafCNode._

/**
 * @author gjoranv
 */

object BuilderGenerator {

  def getBuilder(node: InnerCNode): String = {
    getDeclaration(node) + "\n" +
      indentCode(Indentation,
        getUninitializedScalars(node) + "\n\n" +
          node.getChildren.map(getBuilderFieldDefinition).mkString("\n") + "\n\n" +
          getBuilderConstructors(node, nodeClass(node)) + "\n\n" +
          getOverrideMethod(node) + "\n\n" +
          getBuilderSetters(node) + "\n" +
          getSpecialRootBuilderCode(node)
      ) +
      "}"
  }

  private def getDeclaration(node: InnerCNode) = {
    def getInterfaces =
      if (node.getParent == null) "implements ConfigInstance.Builder"
      else "implements ConfigBuilder"

    "public static class Builder " + getInterfaces + " {"
  }

  private def getSpecialRootBuilderCode(node: InnerCNode) = {
    if (node.getParent == null) "\n" + getDispatchCode(node) + "\n"
    else ""
  }

  private def getDispatchCode(node: InnerCNode) = {
    // Use full path to @Override, as users are free to define an inner node called 'override'. (summarymap.def does)
    // The generated inner 'Override' class would otherwise be mistaken for the annotation.
    """
      |@java.lang.Override
      |public final boolean dispatchGetConfig(ConfigInstance.Producer producer) {
      |  if (producer instanceof Producer) {
      |    ((Producer)producer).getConfig(this);
      |    return true;
      |  }
      |  return false;
      |}
      |
      |@java.lang.Override
      |public final String getDefMd5()       { return CONFIG_DEF_MD5; }
      |@java.lang.Override
      |public final String getDefName()      { return CONFIG_DEF_NAME; }
      |@java.lang.Override
      |public final String getDefNamespace() { return CONFIG_DEF_NAMESPACE; }
    """.stripMargin.trim
  }

  private def getUninitializedScalars(node: InnerCNode): String = {
    val scalarsWithoutDefault = {
      node.getChildren.collect {
        case leaf: LeafCNode if (!leaf.isArray && !leaf.isMap && leaf.getDefaultValue == null) =>
          "\"" + leaf.getName + "\""
      }
    }

    val uninitializedList =
      if (scalarsWithoutDefault.size > 0)
        "Arrays.asList(\n" + indentCode(Indentation, scalarsWithoutDefault.mkString("",",\n","\n)"))
      else
        ""

    "private Set<String> " + InternalPrefix + "uninitialized = new HashSet<String>(" + uninitializedList + ");"
  }

  private def getBuilderFieldDefinition(node: CNode): String = {

    (node match {
      case array if node.isArray =>
        "public List<%s> %s = new ArrayList<>()".format(builderType(array), array.getName)
      case map if node.isMap =>
        "public Map<String, %s> %s = new LinkedHashMap<>()".format(builderType(map), map.getName)
      case struct: InnerCNode =>
        "public %s %s = new %s()".format(builderType(struct), struct.getName, builderType(struct))
      case scalar : LeafCNode =>
        "private " + boxedBuilderType(scalar) + " " + scalar.getName + " = null"
    }) + ";"
  }

  private def getBuilderSetters(node: CNode): String = {
    val children: Array[CNode] = node.getChildren

    def structSetter(node: InnerCNode) = {
      <code>
        |public Builder {node.getName}({builderType(node)} {InternalPrefix}builder) {{
        |  {node.getName} = {InternalPrefix}builder;
        |  return this;
        |}}
      </code>.text.stripMargin.trim
    }

    def innerArraySetters(node: InnerCNode) = {
      <code>
        |/**
        | * Add the given builder to this builder's list of {nodeClass(node)} builders
        | * @param {InternalPrefix}builder a builder
        | * @return this builder
        | */
        |public Builder {node.getName}({builderType(node)} {InternalPrefix}builder) {{
        |  {node.getName}.add({InternalPrefix}builder);
        |  return this;
        |}}
        |
        |/**
        | * Set the given list as this builder's list of {nodeClass(node)} builders
        | * @param __builders a list of builders
        | * @return this builder
        | */
        |public Builder {node.getName}(List&lt;{builderType(node)}&gt; __builders) {{
        |  {node.getName} = __builders;
        |  return this;
        |}}
      </code>.text.stripMargin.trim
    }

    def leafArraySetters(node: LeafCNode) = {
      val setters =
      <code>
        |public Builder {node.getName}({builderType(node)} {InternalPrefix}value) {{
        |  {node.getName}.add({InternalPrefix}value);
        |  return this;
        |}}
        |
        |public Builder {node.getName}(Collection&lt;{builderType(node)}&gt; {InternalPrefix}values) {{
        |  {node.getName}.addAll({InternalPrefix}values);
        |  return this;
        |}}
      </code>.text.stripMargin.trim

      val privateSetter =
        if (builderType(node) == "String" || builderType(node) == "FileReference")
          ""
        else
          "\n\n" +
          <code>
          |
          |
          |private Builder {node.getName}(String {InternalPrefix}value) {{
          |  return {node.getName}({builderType(node)}.valueOf({InternalPrefix}value));
          |}}
          </code>.text.stripMargin.trim

       setters + privateSetter
    }

    def innerMapSetters(node: CNode) = {
      <code>
        |public Builder {node.getName}(String {InternalPrefix}key, {builderType(node)} {InternalPrefix}value) {{
        |  {node.getName}.put({InternalPrefix}key, {InternalPrefix}value);
        |  return this;
        |}}
        |
        |public Builder {node.getName}(Map&lt;String, {builderType(node)}&gt; {InternalPrefix}values) {{
        |  {node.getName}.putAll({InternalPrefix}values);
        |  return this;
        |}}
      </code>.text.stripMargin.trim
    }

    def leafMapSetters(node: LeafCNode) = {
      val privateSetter =
        if (builderType(node) == "String" || builderType(node) == "FileReference")
          ""
        else
          "\n\n" +
          <code>
          |
          |
          |private Builder {node.getName}(String {InternalPrefix}key, String {InternalPrefix}value) {{
          |  return {node.getName}({InternalPrefix}key, {builderType(node)}.valueOf({InternalPrefix}value));
          |}}
          </code>.text.stripMargin.trim

       innerMapSetters(node) + privateSetter
    }

    def scalarSetters(node: LeafCNode): String = {
      val name = node.getName

      val signalInitialized =
        if (node.getDefaultValue == null) InternalPrefix + "uninitialized.remove(\"" + name + "\");\n"
        else ""

      val stringSetter =
        builderType(node) match {
          case "String"        => ""
          case "FileReference" => ""
          case _ =>
            """|
              |private Builder %s(String %svalue) {
              |  return %s(%s.valueOf(%svalue));
              |}""".stripMargin.format(name, InternalPrefix,
              name, boxedDataType(node), InternalPrefix)
        }

      def getNullGuard = {
        if (builderType(node) != boxedBuilderType(node))
          ""
        else
          "\n" + "if (%svalue == null) throw new IllegalArgumentException(\"Null value is not allowed.\");"
            .format(InternalPrefix)
      }

      // TODO: check if 2.9.2 allows string to start with a newline
      """|public Builder %s(%s %svalue) {%s
         |  %s = %svalue;
         |  %s
      """.stripMargin.format(name, builderType(node), InternalPrefix, getNullGuard,
        name, InternalPrefix,
        signalInitialized).trim +
        "\n  return this;" + "\n}\n" +
        stringSetter
    }

    (children collect {
      case innerArray: InnerCNode if innerArray.isArray => innerArraySetters(innerArray)
      case innerMap: InnerCNode if innerMap.isMap       => innerMapSetters(innerMap)
      case leafArray: LeafCNode if leafArray.isArray    => leafArraySetters(leafArray)
      case leafMap: LeafCNode if leafMap.isMap          => leafMapSetters(leafMap)
      case struct: InnerCNode => structSetter(struct)
      case scalar: LeafCNode  => scalarSetters(scalar)
    } ).mkString("\n\n")
  }

  private def getBuilderConstructors(node: CNode, className: String): String = {
    def setBuilderValueFromConfig(child: CNode) = {
      val name = child.getName
      val isArray = child.isArray
      val isMap = child.isMap

      child match {
        case fileArray: FileLeaf if isArray => name + "(" + userDataType(fileArray) + ".toValues(config." + name + "()));"
        case fileMap: FileLeaf if isMap => name + "(" + userDataType(fileMap) + ".toValueMap(config." + name + "()));"
        case file: FileLeaf => name + "(config." + name + "().value());"
        case pathArray: PathLeaf if isArray => name + "(" + nodeClass(pathArray) + ".toFileReferences(config." + name + "));"
        case pathMap: PathLeaf if isMap => name + "(" + nodeClass(pathMap) + ".toFileReferenceMap(config." + name + "));"
        case path: PathLeaf => name + "(config." + name + ".getFileReference());"
        case leaf: LeafCNode => name + "(config." + name + "());"
        case innerArray: InnerCNode if isArray => setInnerArrayBuildersFromConfig(innerArray)
        case innerMap: InnerCNode   if isMap   => setInnerMapBuildersFromConfig(innerMap)
        case struct => name + "(new " +  builderType(struct) + "(config." + name + "()));"
      }
    }

    def setInnerArrayBuildersFromConfig(innerArr: InnerCNode) = {
      val elemName = createUniqueSymbol(node, innerArr.getName)
      <code>
        |for ({userDataType(innerArr)} {elemName} : config.{innerArr.getName}()) {{
        |  {innerArr.getName}(new {builderType(innerArr)}({elemName}));
        |}}
      </code>.text.stripMargin.trim
    }

    def setInnerMapBuildersFromConfig(innerMap: InnerCNode) = {
      val entryName = InternalPrefix + "entry"
      <code>
        |for (Map.Entry&lt;String, {userDataType(innerMap)}&gt; {entryName} : config.{innerMap.getName}().entrySet()) {{
        |  {innerMap.getName}({entryName}.getKey(), new {userDataType(innerMap)}.Builder({entryName}.getValue()));
        |}}
      </code>.text.stripMargin.trim
    }

    <code>
      |public Builder() {{ }}
      |
      |public Builder({className} config) {{
      |{indentCode(Indentation, node.getChildren.map(setBuilderValueFromConfig).mkString("\n"))}
      |}}
    </code>.text.stripMargin.trim
  }

  def arrayOverride(name: String, superior: String): String = {
    Indentation + name + ".addAll(" + superior + "." + name + ");"
  }

  private def getOverrideMethod(node:CNode): String =  {
    val method = "override"
    val superior = InternalPrefix + "superior"

    def callSetter(name: String): String = {
      name + "(" + superior + "." + name + ");"
    }
    def overrideBuilderValue(child: CNode) = {
      val name = child.getName
      child match {
        case leafArray: CNode if (child.isArray) =>
          conditionStatement(child) + "\n" + arrayOverride(name, superior)
        case struct: InnerCNode if !(child.isArray || child.isMap) =>
          name + "(" + name + "." + method + "(" + superior + "." + name + "));"
        case map: CNode if child.isMap =>
          callSetter(name)
        case _ =>
          conditionStatement(child) + "\n" +
            Indentation + callSetter(name)
      }
    }

    def conditionStatement(child: CNode) = {
      val name = child.getName
      val isArray = child.isArray
      val isMap = child.isMap
      child match {
        case _ if isArray => "if (!" + superior + "." + name + ".isEmpty())"
        case _ if isMap => ""
        case scalar: LeafCNode => "if (" + superior + "." + name + " != null)"
        case struct => ""
      }
    }

    <code>
      |private Builder {method}(Builder {superior}) {{
      |{indentCode(Indentation, node.getChildren.map(overrideBuilderValue).mkString("\n"))}
      |  return this;
      |}}
    </code>.text.stripMargin.trim
  }

  private def builderType(node: CNode): String = {
    node match {
      case inner: InnerCNode => boxedDataType(node) + ".Builder"
      case file: FileLeaf => "String"
      case path: PathLeaf => "FileReference"
      case leafArray: LeafCNode if (node.isArray || node.isMap) => boxedDataType(node)
      case _ => userDataType(node)
    }
  }

  private def boxedBuilderType(node: LeafCNode): String = {
    node match {
      case file: FileLeaf => "String"
      case path: PathLeaf => "FileReference"
      case _ => boxedDataType(node)
    }
  }

}

File: configgen/src/main/java/com/yahoo/config/codegen/BuilderGenerator.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.config.codegen;

import com.yahoo.config.codegen.LeafCNode.FileLeaf;
import com.yahoo.config.codegen.LeafCNode.ModelLeaf;
import com.yahoo.config.codegen.LeafCNode.PathLeaf;
import com.yahoo.config.codegen.LeafCNode.OptionalPathLeaf;
import com.yahoo.config.codegen.LeafCNode.UrlLeaf;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import static com.yahoo.config.codegen.ConfigGenerator.boxedDataType;
import static com.yahoo.config.codegen.ConfigGenerator.indentCode;
import static com.yahoo.config.codegen.ConfigGenerator.nodeClass;
import static com.yahoo.config.codegen.ConfigGenerator.userDataType;
import static com.yahoo.config.codegen.JavaClassBuilder.INDENTATION;
import static com.yahoo.config.codegen.JavaClassBuilder.createUniqueSymbol;
import static com.yahoo.config.codegen.ReservedWords.INTERNAL_PREFIX;
import static java.util.Arrays.stream;

/**
 * @author gjoranv
 * @author ollivir
 */
public class BuilderGenerator {

    public static String getBuilder(InnerCNode node) {
        return getDeclaration(node) + "\n" + //
                indentCode(INDENTATION, getUninitializedScalars(node) + "\n\n" + //
                        stream(node.getChildren()).map(BuilderGenerator::getBuilderFieldDefinition).collect(Collectors.joining("\n"))
                        + "\n\n" + //
                        getBuilderConstructors(node, nodeClass(node)) + "\n\n" + //
                        getOverrideMethod(node) + "\n\n" + //
                        getBuilderSetters(node) + "\n" + //
                        getSpecialRootBuilderCode(node) + "\n" + //
                        getBuildMethod(node) + "\n") //
                + "}";
    }

    private static String getDeclaration(InnerCNode node) {
        String getInterfaces = (node.getParent() == null) ? "implements ConfigInstance.Builder" : "implements ConfigBuilder";

        return "public static final class Builder " + getInterfaces + " {";
    }

    private static String getSpecialRootBuilderCode(InnerCNode node) {
        return (node.getParent() == null) ? "\n" + getRootDeclarations() + "\n" : "";
    }

    private static String getBuildMethod(InnerCNode node) {
        return "public " + nodeClass(node) + " build() {\n" +
               "  return new " + nodeClass(node) + "(this);\n" +
               "}\n";
    }

    private static String getRootDeclarations() {
        // Use full path to @Override, as users are free to define an inner node called
        // 'override'. (summarymap.def does)
        // The generated inner 'Override' class would otherwise be mistaken for the
        // annotation.
        return  "private boolean _applyOnRestart = false;\n" +
                "\n" +
                "@java.lang.Override\n" +
                "public final boolean dispatchGetConfig(ConfigInstance.Producer producer) {\n" +
                "  if (producer instanceof Producer) {\n" +
                "    ((Producer)producer).getConfig(this);\n" +
                "    return true;\n" +
                "  }\n" + //
                "  return false;\n" +
                "}\n" +
                "\n" +
                "@java.lang.Override\n" +
                "public final String getDefMd5() { return CONFIG_DEF_MD5; }\n" +
                "\n" +
                "@java.lang.Override\n" +
                "public final String getDefName() { return CONFIG_DEF_NAME; }\n" +
                "\n" +
                "@java.lang.Override\n" +
                "public final String getDefNamespace() { return CONFIG_DEF_NAMESPACE; }\n" +
                "\n" +
                "@java.lang.Override\n" +
                "public final boolean getApplyOnRestart() { return _applyOnRestart; }\n" +
                "\n" +
                "@java.lang.Override\n" +
                "public final void setApplyOnRestart(boolean applyOnRestart) { _applyOnRestart = applyOnRestart; }";
    }

    private static String getUninitializedScalars(InnerCNode node) {
        List<String> scalarsWithoutDefault = new ArrayList<>();
        for (CNode child : node.getChildren()) {
            if (child instanceof LeafCNode
                    && (!child.isArray && !child.isMap && ((LeafCNode) child).getDefaultValue() == null)
                    && (! (child instanceof OptionalPathLeaf))) {
                scalarsWithoutDefault.add("\"" + child.getName() + "\"");
            }
        }

        String uninitializedList = (scalarsWithoutDefault.size() > 0)
                ? "List.of(\n" + indentCode(INDENTATION, String.join(",\n", scalarsWithoutDefault) + "\n)")
                : "";

        return "private Set<String> " + INTERNAL_PREFIX + "uninitialized = new HashSet<String>(" + uninitializedList + ");";
    }

    private static String getBuilderFieldDefinition(CNode node) {
        if (node.isArray) {
            return String.format("public List<%s> %s = new ArrayList<>();", builderType(node), node.getName());
        } else if (node.isMap) {
            return String.format("public Map<String, %s> %s = new LinkedHashMap<>();", builderType(node), node.getName());
        } else if (node instanceof InnerCNode) {
            return String.format("public %s %s = new %s();", builderType(node), node.getName(), builderType(node));
        } else if (node instanceof LeafCNode) {
            String boxedBuilderType = boxedBuilderType((LeafCNode) node);
            if (boxedBuilderType.startsWith("Optional<"))
                return String.format("private %s %s = Optional.empty();", boxedBuilderType, node.getName());
            else
                return String.format("private %s %s = null;", boxedBuilderType, node.getName());
        } else {
            throw new IllegalStateException("Cannot produce builder field definition for node"); // Should not happen
        }
    }

    private static String getBuilderSetters(CNode node) {
        List<String> elem = new ArrayList<>();
        CNode[] children = node.getChildren();

        for (CNode child : children) {
            if (child instanceof InnerCNode && child.isArray) {
                elem.add(BuilderSetters.innerArraySetters((InnerCNode) child));
            } else if (child instanceof InnerCNode && child.isMap) {
                elem.add(BuilderSetters.innerMapSetters(child));
            } else if (child instanceof LeafCNode && child.isArray) {
                elem.add(BuilderSetters.leafArraySetters((LeafCNode) child));
            } else if (child instanceof LeafCNode && child.isMap) {
                elem.add(BuilderSetters.leafMapSetters(child));
            } else if (child instanceof InnerCNode) {
                elem.add(BuilderSetters.structSetter((InnerCNode) child));
            } else if (child instanceof LeafCNode) {
                elem.add(BuilderSetters.scalarSetters((LeafCNode) child));
            }
        }
        return String.join("\n\n", elem);
    }

    private static class BuilderSetters {

        private static String structSetter(InnerCNode n) {
            return "public Builder " + n.getName() + "(" + builderType(n) + " " + INTERNAL_PREFIX + "builder) {\n" + //
                    "  " + n.getName() + " = " + INTERNAL_PREFIX + "builder;\n" + //
                    "  return this;\n" + //
                    "}\n" + //
                    "/**\n" + //
                    " * Make a new builder and run the supplied function on it before adding it to the list\n" + //
                    " * @param __func lambda that modifies the given builder\n" + //
                    " * @return this builder\n" + //
                    " */\n" + //
                    "public Builder " + n.getName() + "(java.util.function.Consumer<" + builderType(n) + "> __func) {\n" + //
                    "  " + builderType(n) + " __inner = new " + builderType(n) +"();\n" + //
                    "  __func.accept(__inner);\n" + //
                    "  " + n.getName() + " = __inner;\n" + //
                    "  return this;\n" + //
                    "}";

        }

        private static String innerArraySetters(InnerCNode n) {
            return "/**\n" + //
                    " * Add the given builder to this builder's list of " + nodeClass(n) + " builders\n" + //
                    " * @param " + INTERNAL_PREFIX + "builder a builder\n" + //
                    " * @return this builder\n" + //
                    " */\n" + //
                    "public Builder " + n.getName() + "(" + builderType(n) + " " + INTERNAL_PREFIX + "builder) {\n" + //
                    "  " + n.getName() + ".add(" + INTERNAL_PREFIX + "builder);\n" + //
                    "  return this;\n" + //
                    "}\n" + //
                    "\n" + //
                    "/**\n" + //
                    " * Make a new builder and run the supplied function on it before adding it to the list\n" + //
                    " * @param __func lambda that modifies the given builder\n" + //
                    " * @return this builder\n" + //
                    " */\n" + //
                    "public Builder " + n.getName() + "(java.util.function.Consumer<" + builderType(n) + "> __func) {\n" + //
                    "  " + builderType(n) + " __inner = new " + builderType(n) +"();\n" + //
                    "  __func.accept(__inner);\n" + //
                    "  " + n.getName() + ".add(__inner);\n" + //
                    "  return this;\n" + //
                    "}\n" + //
                    "\n" + //
                    "/**\n" + //
                    " * Set the given list as this builder's list of " + nodeClass(n) + " builders\n" + //
                    " * @param __builders a list of builders\n" + //
                    " * @return this builder\n" + //
                    " */\n" + //
                    "public Builder " + n.getName() + "(List<" + builderType(n) + "> __builders) {\n" + //
                    "  " + n.getName() + " = __builders;\n" + //
                    "  return this;\n" + //
                    "}";
        }

        private static String publicLeafNodeSetters(LeafCNode n) {
            return "public Builder " + n.getName() + "(" + builderType(n) + " " + INTERNAL_PREFIX + "value) {\n" + //
                    "  " + n.getName() + ".add(" + INTERNAL_PREFIX + "value);\n" + //
                    "  return this;\n" + //
                    "}\n" + //
                    "\n" + //
                    "public Builder " + n.getName() + "(Collection<" + builderType(n) + "> " + INTERNAL_PREFIX + "values) {\n" + //
                    "  " + n.getName() + ".addAll(" + INTERNAL_PREFIX + "values);\n" + //
                    "  return this;\n" + //
                    "}";
        }

        private static String privateLeafNodeSetter(LeafCNode n) {
            if ("String".equals(builderType(n)) || "FileReference".equals(builderType(n))) {
                return "";
            } else if ("Optional<FileReference>".equals(builderType(n))) {
                return "\n\n" + //
                        "private Builder " + n.getName() + "(String " + INTERNAL_PREFIX + "value) {\n" + //
                        "  return " + n.getName() + "(" + builderType(n) + ".of(" + INTERNAL_PREFIX + "value));\n" + //
                        "}";
            } else {
                return "\n\n" + //
                        "private Builder " + n.getName() + "(String " + INTERNAL_PREFIX + "value) {\n" + //
                        "  return " + n.getName() + "(" + builderType(n) + ".valueOf(" + INTERNAL_PREFIX + "value));\n" + //
                        "}";
            }
        }

        private static String leafArraySetters(LeafCNode n) {
            return publicLeafNodeSetters(n) + privateLeafNodeSetter(n);
        }

        private static String innerMapSetters(CNode n) {
            String r = "public Builder " + n.getName() + "(String " + INTERNAL_PREFIX + "key, " + builderType(n) + " " + INTERNAL_PREFIX + "value) {\n" + //
                    "  " + n.getName() + ".put(" + INTERNAL_PREFIX + "key, " + INTERNAL_PREFIX + "value);\n" + //
                    "  return this;\n" + //
                    "}\n" + //
                    "\n" + //
                    "public Builder " + n.getName() + "(Map<String, " + builderType(n) + "> " + INTERNAL_PREFIX + "values) {\n" + //
                    "  " + n.getName() + ".putAll(" + INTERNAL_PREFIX + "values);\n" + //
                    "  return this;\n" + //
                    "}";
            if (n instanceof InnerCNode) {
                r = r +
                    "\n\n" + //
                    "/**\n" + //
                    " * Make a new builder and run the supplied function on it before using it as the value\n" + //
                    " * @param __func lambda that modifies the given builder\n" + //
                    " * @return this builder\n" + //
                    " */\n" + //
                    "public Builder " + n.getName() + "(String __key, java.util.function.Consumer<" + builderType(n) + "> __func) {\n" + //
                    "  " + builderType(n) + " __inner = new " + builderType(n) +"();\n" + //
                    "  __func.accept(__inner);\n" + //
                    "  " + n.getName() + ".put(__key, __inner);\n" + //
                    "  return this;\n" + //
                    "}";
            }
            return r;
        }

        private static String privateLeafMapSetter(CNode n) {
            if ("String".equals(builderType(n)) || "FileReference".equals(builderType(n))) {
                return "";
            } else {
                return "\n\n" + //
                        "private Builder " + n.getName() + "(String " + INTERNAL_PREFIX + "key, String " + INTERNAL_PREFIX + "value) {\n" + //
                        "  return " + n.getName() + "(" + INTERNAL_PREFIX + "key, " + builderType(n) + ".valueOf(" + INTERNAL_PREFIX
                        + "value));\n" + //
                        "}";
            }
        }

        private static String leafMapSetters(CNode n) {
            return innerMapSetters(n) + privateLeafMapSetter(n);
        }

        private static String scalarSetters(LeafCNode n) {
            String name = n.getName();

            String signalInitialized = (n.getDefaultValue() == null) ? "  " + INTERNAL_PREFIX + "uninitialized.remove(\"" + name + "\");\n"
                    : "";

            String bType = builderType(n);
            String privateSetter = "";
            if ( ! Set.of("String", "FileReference", "ModelReference", "Optional<FileReference>").contains(bType)) {
                String type = boxedDataType(n);
                if ("UrlReference".equals(bType))
                    type = bType;
                //
                System.err.println(String.format("gen privateSetter for '%s' bType '%s'", name, bType));
                privateSetter = String.format("""

                                                      private Builder %s(String %svalue) {
                                                        return %s(%s.valueOf(%svalue));
                                                      }""", name, INTERNAL_PREFIX, name, type, INTERNAL_PREFIX);
            } else if ("Optional<FileReference>".equals(bType)) {
                //
                privateSetter = String.format("""

                                                      private Builder %s(FileReference %svalue) {
                                                        return %s(Optional.of(%svalue));
                                                      }""", name, INTERNAL_PREFIX, name, INTERNAL_PREFIX);
            }

            String getNullGuard = bType.equals(boxedBuilderType(n)) ? String.format(
                    "\nif (%svalue == null) throw new IllegalArgumentException(\"Null value is not allowed.\");", INTERNAL_PREFIX) : "";

            return String.format("public Builder %s(%s %svalue) {%s\n" +
                    "  %s = %svalue;\n" + //
                    "%s", name, bType, INTERNAL_PREFIX, getNullGuard, name, INTERNAL_PREFIX, signalInitialized) +
                    "  return this;" + "\n}\n" + privateSetter;
        }
    }

    private static String setBuilderValueFromConfig(CNode child, CNode node) {
        String name = child.getName();
        boolean isArray = child.isArray;
        boolean isMap = child.isMap;

        if (child instanceof FileLeaf && isArray) {
            return name + "(" + userDataType(child) + ".toValues(config." + name + "()));";
        } else if (child instanceof FileLeaf && isMap) {
            return name + "(" + userDataType(child) + ".toValueMap(config." + name + "()));";
        } else if (child instanceof FileLeaf) {
            return name + "(config." + name + "().value());";
        } else if (child instanceof PathLeaf && isArray) {
            return name + "(" + nodeClass(child) + ".toFileReferences(config." + name + "));";
        } else if (child instanceof PathLeaf && isMap) {
            return name + "(" + nodeClass(child) + ".toFileReferenceMap(config." + name + "));";
        } else if (child instanceof PathLeaf) {
            return name + "(config." + name + ".getFileReference());";
        } else if (child instanceof OptionalPathLeaf) {
            return name + "(config." + name + ".getFileReference());";
        } else if (child instanceof UrlLeaf && isArray) {
            return name + "(" + nodeClass(child) + ".toUrlReferences(config." + name + "));";
        } else if (child instanceof UrlLeaf && isMap) {
            return name + "(" + nodeClass(child) + ".toUrlReferenceMap(config." + name + "));";
        } else if (child instanceof UrlLeaf) {
            return name + "(config." + name + ".getUrlReference());";
        } else if (child instanceof ModelLeaf && isArray) {
            return name + "(" + nodeClass(child) + ".toModelReferences(config." + name + "));";
        } else if (child instanceof ModelLeaf && isMap) {
            return name + "(" + nodeClass(child) + ".toModelReferenceMap(config." + name + "));";
        } else if (child instanceof ModelLeaf) {
            return name + "(config." + name + ".getModelReference());";
        } else if (child instanceof LeafCNode) {
            return name + "(config." + name + "());";
        } else if (child instanceof InnerCNode && isArray) {
            return setInnerArrayBuildersFromConfig((InnerCNode) child, node);
        } else if (child instanceof InnerCNode && isMap) {
            return setInnerMapBuildersFromConfig((InnerCNode) child);
        } else {
            return name + "(new " + builderType(child) + "(config." + name + "()));";
        }
    }

    private static String setInnerArrayBuildersFromConfig(InnerCNode innerArr, CNode node) {
        String elemName = createUniqueSymbol(node, innerArr.getName());

        return "for (" + userDataType(innerArr) + " " + elemName + " : config." + innerArr.getName() + "()) {\n" + //
                "  " + innerArr.getName() + "(new " + builderType(innerArr) + "(" + elemName + "));\n" + //
                "}";
    }

    private static String setInnerMapBuildersFromConfig(InnerCNode innerMap) {
        String entryName = INTERNAL_PREFIX + "entry";
        return "for (Map.Entry<String, " + userDataType(innerMap) + "> " + entryName + " : config." + innerMap.getName()
                + "().entrySet()) {\n" + //
                "  " + innerMap.getName() + "(" + entryName + ".getKey(), new " + userDataType(innerMap) + ".Builder(" + entryName
                + ".getValue()));\n" + //
                "}";
    }

    private static String getBuilderConstructors(CNode node, String className) {
        return "public Builder() { }\n" + //
                "\n" + //
                "public Builder(" + className + " config) {\n" + //
                indentCode(INDENTATION,
                        stream(node.getChildren()).map(child -> setBuilderValueFromConfig(child, node)).collect(Collectors.joining("\n")))
                + //
                "\n}";
    }

    private static String conditionStatement(CNode child) {
        String superior = INTERNAL_PREFIX + "superior";

        if (child.isArray) {
            return "if (!" + superior + "." + child.getName() + ".isEmpty())";
        } else if (child.isMap) {
            return "";
        } else if (child instanceof LeafCNode) {
            return "if (" + superior + "." + child.getName() + " != null)";
        } else {
            return "";
        }
    }

    private static String overrideBuilderValue(CNode child) {
        final String superior = INTERNAL_PREFIX + "superior";
        final String method = "override";
        final String name = child.getName();
        final String callSetter = name + "(" + superior + "." + name + ");";

        if (child.isArray) {
            String arrayOverride = INDENTATION + name + ".addAll(" + superior + "." + name + ");";
            return conditionStatement(child) + "\n" + arrayOverride;
        } else if (child instanceof InnerCNode && !child.isArray && !child.isMap) {
            return name + "(" + name + "." + method + "(" + superior + "." + name + "));";
        } else if (child.isMap) {
            return callSetter;
        } else {
            return conditionStatement(child) + "\n" + INDENTATION + callSetter;
        }
    }

    private static String getOverrideMethod(CNode node) {
        String superior = INTERNAL_PREFIX + "superior";
        String method = "override";

        return "private Builder " + method + "(Builder " + superior + ") {\n" + //
                indentCode(INDENTATION,
                        stream(node.getChildren()).map(BuilderGenerator::overrideBuilderValue).collect(Collectors.joining("\n")))
                + "\n" + //
                "  return this;\n" + //
                "}";
    }

    private static String builderType(CNode node) {
        if (node instanceof InnerCNode) {
            return boxedDataType(node) + ".Builder";
        } else if (node instanceof FileLeaf) {
            return "String";
        } else if (node instanceof PathLeaf) {
            return "FileReference";
        } else if (node instanceof OptionalPathLeaf) {
            return "Optional<FileReference>";
        } else if (node instanceof UrlLeaf) {
            return "UrlReference";
        } else if (node instanceof ModelLeaf) {
            return "ModelReference";
        } else if (node instanceof LeafCNode && (node.isArray || node.isMap)) {
            return boxedDataType(node);
        } else {
            return userDataType(node);
        }
    }

    private static String boxedBuilderType(LeafCNode node) {
        if (node instanceof FileLeaf) {
            return "String";
        } else if (node instanceof PathLeaf) {
            return "FileReference";
        } else if (node instanceof OptionalPathLeaf) {
            return "Optional<FileReference>";
        } else if (node instanceof UrlLeaf) {
            return "UrlReference";
        } else if (node instanceof ModelLeaf) {
            return "ModelReference";
        } else {
            return boxedDataType(node);
        }
    }

}
